### PR TITLE
Add weakening_map

### PR DESCRIPTION
**Pull request by AntoinePrv**
A type of map that weakens key to find a match.
The generic implementation of `AuthenticationDataBase`, also needed for `custom_channels`.

### CODE REVIEW

File: libmamba/include/mamba/util/weakening_map.hpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        using typename Base::mapped_type;
+        using typename Base::value_type;
+        using typename Base::size_type;
+        using typename Base::iterator;
+        using typename Base::const_iterator;
+        using weakener_type = Weakener;
+
+        using Base::Base;
+        template <typename... Args>
+        weakening_map(weakener_type weakener, Args&&... args);
```
Comment (JohanMabille) [inline libmamba/include/mamba/util/weakening_map.hpp:38]: Should be declared as explicit

File: libmamba/include/mamba/util/weakening_map.hpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        using Base::swap;
+        using Base::extract;
+        using Base::merge;
+
+        using Base::reserve;
+
+        using Base::at;
+
+        [[nodiscard]] auto at_weaken(const key_type& key) -> mapped_type&;
+        [[nodiscard]] auto at_weaken(const key_type& key) const -> const mapped_type&;
```
Comment (JohanMabille) [inline libmamba/include/mamba/util/weakening_map.hpp:65]: Should this replace the `at` API? Asked differently, does the user needs both methods? Is there a need to distinguish between strict finding and weaken finding?

File: libmamba/include/mamba/util/weakening_map.hpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+        using Base::find;
+
+        auto find_weaken(const key_type& key) -> iterator;
+        auto find_weaken(const key_type& key) const -> const_iterator;
```
Comment (JohanMabille) [inline libmamba/include/mamba/util/weakening_map.hpp:70]: Same question here.

File: libmamba/include/mamba/util/weakening_map.hpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+        [[nodiscard]] auto contains(const key_type& key) const -> bool;
+
+        [[nodiscard]] auto contains_weaken(const key_type& key) const -> bool;
```
Comment (JohanMabille) [inline libmamba/include/mamba/util/weakening_map.hpp:74]: And here.

File: libmamba/include/mamba/util/weakening_map.hpp (hunk 2 above)
Comment (AntoinePrv) [reply libmamba/include/mamba/util/weakening_map.hpp:69]: I'd say it's still relevant, for instance if users want to remove an entry, they would more likely want an exact match.

**Review by JohanMabille — APPROVED** [review]

**Pull request merged**